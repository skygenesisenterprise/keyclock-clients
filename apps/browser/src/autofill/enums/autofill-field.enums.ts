export const AutofillFieldQualifier = {
  password: "password",
  newPassword: "newPassword",
  username: "username",
  cardholderName: "cardholderName",
  cardNumber: "cardNumber",
  cardExpirationMonth: "cardExpirationMonth",
  cardExpirationYear: "cardExpirationYear",
  cardExpirationDate: "cardExpirationDate",
  cardCvv: "cardCvv",
  identityTitle: "identityTitle",
  identityFirstName: "identityFirstName",
  identityMiddleName: "identityMiddleName",
  identityLastName: "identityLastName",
  identityFullName: "identityFullName",
  identityAddress1: "identityAddress1",
  identityAddress2: "identityAddress2",
  identityAddress3: "identityAddress3",
  identityCity: "identityCity",
  identityState: "identityState",
  identityPostalCode: "identityPostalCode",
  identityCountry: "identityCountry",
  identityCompany: "identityCompany",
  identityPhone: "identityPhone",
  identityEmail: "identityEmail",
  identityUsername: "identityUsername",
} as const;

export type AutofillFieldQualifierType =
  (typeof AutofillFieldQualifier)[keyof typeof AutofillFieldQualifier];
