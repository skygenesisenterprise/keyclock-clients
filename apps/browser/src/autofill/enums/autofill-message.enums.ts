export const AutofillMessageCommand = {
  collectPageDetails: "collectPageDetails",
  collectPageDetailsResponse: "collectPageDetailsResponse",
} as const;

export type AutofillMessageCommandType =
  (typeof AutofillMessageCommand)[keyof typeof AutofillMessageCommand];

export const AutofillMessageSender = {
  collectPageDetailsFromTabObservable: "collectPageDetailsFromTabObservable",
} as const;

export type AutofillMessageSenderType =
  (typeof AutofillMessageSender)[keyof typeof AutofillMessageSender];
