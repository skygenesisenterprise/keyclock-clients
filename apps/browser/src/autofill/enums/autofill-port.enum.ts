const AutofillPort = {
  InjectedScript: "autofill-injected-script-port",
} as const;

export { AutofillPort };
