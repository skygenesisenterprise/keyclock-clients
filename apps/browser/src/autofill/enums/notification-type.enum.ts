const NotificationType = {
  AddLogin: "add",
  ChangePassword: "change",
  UnlockVault: "unlock",
  AtRiskPassword: "at-risk-password",
} as const;

type NotificationTypes = (typeof NotificationType)[keyof typeof NotificationType];

export { NotificationType, NotificationTypes };
