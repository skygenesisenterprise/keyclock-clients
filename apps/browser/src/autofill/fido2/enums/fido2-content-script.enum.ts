export const Fido2ContentScript = {
  PageScript: "content/fido2-page-script.js",
  PageScriptDelayAppend: "content/fido2-page-script-delay-append-mv2.js",
  ContentScript: "content/fido2-content-script.js",
} as const;

export const Fido2ContentScriptId = {
  PageScript: "fido2-page-script-registration",
  ContentScript: "fido2-content-script-registration",
} as const;
