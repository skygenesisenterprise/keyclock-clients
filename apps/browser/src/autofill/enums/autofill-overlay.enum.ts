import { CipherType } from "@bitwarden/common/vault/enums";

export const AutofillOverlayElement = {
  Button: "autofill-inline-menu-button",
  List: "autofill-inline-menu-list",
} as const;

export type AutofillOverlayElementType =
  (typeof AutofillOverlayElement)[keyof typeof AutofillOverlayElement];

export const AutofillOverlayPort = {
  Button: "autofill-inline-menu-button-port",
  ButtonMessageConnector: "autofill-inline-menu-button-message-connector",
  List: "autofill-inline-menu-list-port",
  ListMessageConnector: "autofill-inline-menu-list-message-connector",
} as const;

export const RedirectFocusDirection = {
  Current: "current",
  Previous: "previous",
  Next: "next",
} as const;

export const InlineMenuFillTypes = {
  AccountCreationUsername: 5,
  PasswordGeneration: 6,
  CurrentPasswordUpdate: 7,
} as const;

export type InlineMenuFillTypeValue =
  (typeof InlineMenuFillTypes)[keyof typeof InlineMenuFillTypes];

export type InlineMenuFillType = InlineMenuFillTypeValue | CipherType;

export const InlineMenuAccountCreationFieldType = {
  Text: "text",
  Email: "email",
  Password: "password",
  Totp: "totp",
} as const;

export type InlineMenuAccountCreationFieldTypes =
  (typeof InlineMenuAccountCreationFieldType)[keyof typeof InlineMenuAccountCreationFieldType];

export const MAX_SUB_FRAME_DEPTH = 8;
