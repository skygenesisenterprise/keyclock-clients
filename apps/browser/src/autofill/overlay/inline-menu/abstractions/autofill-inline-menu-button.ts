import { AuthenticationStatus } from "@bitwarden/common/auth/enums/authentication-status";

export type AutofillInlineMenuButtonMessage = { command: string; colorScheme?: string };

export type UpdateAuthStatusMessage = AutofillInlineMenuButtonMessage & {
  authStatus: AuthenticationStatus;
};

export type InitAutofillInlineMenuButtonMessage = UpdateAuthStatusMessage & {
  styleSheetUrl: string;
  translations: Record<string, string>;
  portKey: string;
};

export type AutofillInlineMenuButtonWindowMessageHandlers = {
  [key: string]: CallableFunction;
  initAutofillInlineMenuButton: ({
    message,
  }: {
    message: InitAutofillInlineMenuButtonMessage;
  }) => void;
  checkAutofillInlineMenuButtonFocused: () => void;
  updateAutofillInlineMenuButtonAuthStatus: ({
    message,
  }: {
    message: UpdateAuthStatusMessage;
  }) => void;
  updateAutofillInlineMenuColorScheme: ({
    message,
  }: {
    message: AutofillInlineMenuButtonMessage;
  }) => void;
};
