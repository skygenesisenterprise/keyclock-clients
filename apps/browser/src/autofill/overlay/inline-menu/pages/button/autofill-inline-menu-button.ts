// FIXME: Update this file to be type safe and remove this and next line
// @ts-strict-ignore
import "@webcomponents/custom-elements";
import "lit/polyfill-support.js";
import { AuthenticationStatus } from "@bitwarden/common/auth/enums/authentication-status";
import { EVENTS } from "@bitwarden/common/autofill/constants";

import { buildSvgDomElement } from "../../../../utils";
import { logoIcon, logoLockedIcon } from "../../../../utils/svg-icons";
import {
  InitAutofillInlineMenuButtonMessage,
  AutofillInlineMenuButtonMessage,
  AutofillInlineMenuButtonWindowMessageHandlers,
} from "../../abstractions/autofill-inline-menu-button";
import { AutofillInlineMenuPageElement } from "../shared/autofill-inline-menu-page-element";

export class AutofillInlineMenuButton extends AutofillInlineMenuPageElement {
  private authStatus: AuthenticationStatus = AuthenticationStatus.LoggedOut;
  private readonly buttonElement: HTMLButtonElement;
  private readonly logoIconElement: HTMLElement;
  private readonly logoLockedIconElement: HTMLElement;
  private readonly inlineMenuButtonWindowMessageHandlers: AutofillInlineMenuButtonWindowMessageHandlers =
    {
      initAutofillInlineMenuButton: ({ message }) => this.initAutofillInlineMenuButton(message),
      checkAutofillInlineMenuButtonFocused: () => this.checkButtonFocused(),
      updateAutofillInlineMenuButtonAuthStatus: ({ message }) =>
        this.updateAuthStatus(message.authStatus),
      updateAutofillInlineMenuColorScheme: ({ message }) => this.updatePageColorScheme(message),
    };

  constructor() {
    super();

    this.buttonElement = globalThis.document.createElement("button");

    this.setupGlobalListeners(this.inlineMenuButtonWindowMessageHandlers);

    this.logoIconElement = buildSvgDomElement(logoIcon);
    this.logoIconElement.classList.add("inline-menu-button-svg-icon", "logo-icon");

    this.logoLockedIconElement = buildSvgDomElement(logoLockedIcon);
    this.logoLockedIconElement.classList.add("inline-menu-button-svg-icon", "logo-locked-icon");
  }

  /**
   * Initializes the inline menu button. Facilitates ensuring that the page
   * is set up with the expected styles and translations.
   *
   * @param authStatus - The authentication status of the user
   * @param styleSheetUrl - The URL of the stylesheet to apply to the page
   * @param translations - The translations to apply to the page
   * @param portKey - Background generated key that allows the port to communicate with the background
   */
  private async initAutofillInlineMenuButton({
    authStatus,
    styleSheetUrl,
    translations,
    portKey,
  }: InitAutofillInlineMenuButtonMessage) {
    const linkElement = await this.initAutofillInlineMenuPage(
      "button",
      styleSheetUrl,
      translations,
      portKey,
    );
    this.buttonElement.tabIndex = -1;
    this.buttonElement.type = "button";
    this.buttonElement.classList.add("inline-menu-button");
    this.buttonElement.setAttribute(
      "aria-label",
      this.getTranslation("toggleBitwardenVaultOverlay"),
    );
    this.buttonElement.addEventListener(EVENTS.CLICK, this.handleButtonElementClick);
    this.postMessageToParent({ command: "updateAutofillInlineMenuColorScheme" });

    this.updateAuthStatus(authStatus);

    this.shadowDom.append(linkElement, this.buttonElement);
  }

  /**
   * Updates the authentication status of the user. This will update the icon
   * displayed on the button.
   *
   * @param authStatus - The authentication status of the user
   */
  private updateAuthStatus(authStatus: AuthenticationStatus) {
    this.authStatus = authStatus;

    this.buttonElement.innerHTML = "";
    const iconElement =
      this.authStatus === AuthenticationStatus.Unlocked
        ? this.logoIconElement
        : this.logoLockedIconElement;
    this.buttonElement.append(iconElement);
  }

  /**
   * Handles updating the page color scheme meta tag. Ensures that the button
   * does not present with a non-transparent background on dark mode pages.
   *
   * @param colorScheme - The color scheme of the iframe's parent page
   */
  private updatePageColorScheme({ colorScheme }: AutofillInlineMenuButtonMessage) {
    const colorSchemeMetaTag = globalThis.document.querySelector("meta[name='color-scheme']");
    colorSchemeMetaTag?.setAttribute("content", colorScheme);
  }

  /**
   * Handles a click event on the button element. Posts a message to the
   * parent window indicating that the button was clicked.
   */
  private handleButtonElementClick = () => {
    this.postMessageToParent({ command: "autofillInlineMenuButtonClicked" });
  };

  /**
   * Checks if the button is focused. If it is not, then it posts a message
   * to the parent window indicating that the inline menu should be closed.
   */
  private checkButtonFocused() {
    if (globalThis.document.hasFocus()) {
      return;
    }

    if (this.isButtonHovered()) {
      globalThis.document.addEventListener(EVENTS.MOUSEOUT, this.handleMouseOutEvent);
      return;
    }

    this.postMessageToParent({ command: "triggerDelayedAutofillInlineMenuClosure" });
  }

  /**
   * Triggers a re-check of the button's focus status when the mouse leaves the button.
   */
  private handleMouseOutEvent = () => {
    globalThis.document.removeEventListener(EVENTS.MOUSEOUT, this.handleMouseOutEvent);
    this.checkButtonFocused();
  };

  /**
   * Identifies whether the button is currently hovered.
   */
  private isButtonHovered() {
    const hoveredElement = this.buttonElement?.querySelector(":hover");
    return !!(
      hoveredElement &&
      (hoveredElement === this.buttonElement || this.buttonElement.contains(hoveredElement))
    );
  }
}
