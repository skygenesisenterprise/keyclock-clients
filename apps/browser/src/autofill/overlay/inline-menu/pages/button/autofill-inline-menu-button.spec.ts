import { AuthenticationStatus } from "@bitwarden/common/auth/enums/authentication-status";

import { createInitAutofillInlineMenuButtonMessageMock } from "../../../../spec/autofill-mocks";
import { flushPromises, postWindowMessage } from "../../../../spec/testing-utils";

import { AutofillInlineMenuButton } from "./autofill-inline-menu-button";

describe("AutofillInlineMenuButton", () => {
  globalThis.customElements.define("autofill-inline-menu-button", AutofillInlineMenuButton);

  let autofillInlineMenuButton: AutofillInlineMenuButton;
  const portKey: string = "inlineMenuButtonPortKey";

  beforeEach(() => {
    document.body.innerHTML = `<autofill-inline-menu-button></autofill-inline-menu-button>`;
    autofillInlineMenuButton = document.querySelector("autofill-inline-menu-button");
    autofillInlineMenuButton["messageOrigin"] = "https://localhost/";
    jest.spyOn(globalThis.document, "createElement");
    jest.spyOn(globalThis.parent, "postMessage");
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe("initAutofillInlineMenuButton", () => {
    it("creates the button element with the locked icon when the user's auth status is not Unlocked", async () => {
      postWindowMessage(
        createInitAutofillInlineMenuButtonMessageMock({
          authStatus: AuthenticationStatus.Locked,
          portKey,
        }),
      );
      await flushPromises();

      expect(autofillInlineMenuButton["buttonElement"]).toMatchSnapshot();
      expect(autofillInlineMenuButton["buttonElement"].querySelector("svg")).toBe(
        autofillInlineMenuButton["logoLockedIconElement"],
      );
    });

    it("creates the button element with the normal icon when the user's auth status is Unlocked ", async () => {
      postWindowMessage(createInitAutofillInlineMenuButtonMessageMock({ portKey }));
      await flushPromises();

      expect(autofillInlineMenuButton["buttonElement"]).toMatchSnapshot();
      expect(autofillInlineMenuButton["buttonElement"].querySelector("svg")).toBe(
        autofillInlineMenuButton["logoIconElement"],
      );
    });

    it("posts a message to the background indicating that the icon was clicked", async () => {
      postWindowMessage(createInitAutofillInlineMenuButtonMessageMock({ portKey }));
      await flushPromises();

      autofillInlineMenuButton["buttonElement"].click();

      expect(globalThis.parent.postMessage).toHaveBeenCalledWith(
        { command: "autofillInlineMenuButtonClicked", portKey },
        "*",
      );
    });
  });

  describe("global event listeners", () => {
    beforeEach(() => {
      postWindowMessage(createInitAutofillInlineMenuButtonMessageMock({ portKey }));
    });

    it("does not post a message to close the autofill inline menu if the element is focused during the focus check", async () => {
      jest.spyOn(globalThis.document, "hasFocus").mockReturnValue(true);

      postWindowMessage({ command: "checkAutofillInlineMenuButtonFocused" });
      await flushPromises();

      expect(globalThis.parent.postMessage).not.toHaveBeenCalledWith({
        command: "triggerDelayedAutofillInlineMenuClosure",
      });
    });

    it("does not post a message to close the autofill inline menu if the button element is hovered", async () => {
      jest.spyOn(globalThis.document, "hasFocus").mockReturnValue(false);
      jest
        .spyOn(autofillInlineMenuButton["buttonElement"], "querySelector")
        .mockReturnValue(autofillInlineMenuButton["buttonElement"]);

      postWindowMessage({ command: "checkAutofillInlineMenuButtonFocused" });
      await flushPromises();

      expect(globalThis.parent.postMessage).not.toHaveBeenCalledWith({
        command: "triggerDelayedAutofillInlineMenuClosure",
      });
    });

    it("triggers a recheck of the button focus state on mouseout", async () => {
      jest.spyOn(globalThis.document, "removeEventListener");
      jest.spyOn(globalThis.document, "hasFocus").mockReturnValue(false);
      jest
        .spyOn(autofillInlineMenuButton["buttonElement"], "querySelector")
        .mockReturnValue(autofillInlineMenuButton["buttonElement"]);
      postWindowMessage({ command: "checkAutofillInlineMenuButtonFocused" });
      await flushPromises();

      globalThis.document.dispatchEvent(new MouseEvent("mouseout"));

      expect(globalThis.document.removeEventListener).toHaveBeenCalledWith(
        "mouseout",
        autofillInlineMenuButton["handleMouseOutEvent"],
      );
    });

    it("posts a message to close the autofill inline menu if the element is not focused during the focus check", async () => {
      jest.spyOn(globalThis.document, "hasFocus").mockReturnValue(false);
      jest.spyOn(autofillInlineMenuButton["buttonElement"], "querySelector").mockReturnValue(null);

      postWindowMessage({ command: "checkAutofillInlineMenuButtonFocused" });
      await flushPromises();

      expect(globalThis.parent.postMessage).toHaveBeenCalledWith(
        { command: "triggerDelayedAutofillInlineMenuClosure", portKey },
        "*",
      );
    });

    it("updates the user's auth status", async () => {
      autofillInlineMenuButton["authStatus"] = AuthenticationStatus.Locked;

      postWindowMessage({
        command: "updateAutofillInlineMenuButtonAuthStatus",
        authStatus: AuthenticationStatus.Unlocked,
      });
      await flushPromises();

      expect(autofillInlineMenuButton["authStatus"]).toBe(AuthenticationStatus.Unlocked);
    });

    it("updates the page color scheme meta tag", async () => {
      const colorSchemeMetaTag = globalThis.document.createElement("meta");
      colorSchemeMetaTag.setAttribute("name", "color-scheme");
      colorSchemeMetaTag.setAttribute("content", "light");
      globalThis.document.head.append(colorSchemeMetaTag);

      postWindowMessage({
        command: "updateAutofillInlineMenuColorScheme",
        colorScheme: "dark",
      });
      await flushPromises();

      expect(colorSchemeMetaTag.getAttribute("content")).toBe("dark");
    });
  });
});
