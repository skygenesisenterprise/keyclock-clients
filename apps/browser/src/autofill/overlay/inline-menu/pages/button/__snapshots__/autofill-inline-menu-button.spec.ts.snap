// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutofillInlineMenuButton initAutofillInlineMenuButton creates the button element with the locked icon when the user's auth status is not Unlocked 1`] = `
<button
  aria-label="toggleBitwardenVaultOverlay"
  class="inline-menu-button"
  tabindex="-1"
  type="button"
>
  <svg
    aria-hidden="true"
    class="inline-menu-button-svg-icon logo-locked-icon"
    fill="none"
    height="16"
    viewBox="0 0 16 16"
    width="16"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      clip-path="url(#a)"
    >
      <path
        d="M12.66.175A.566.566 0 0 0 12.25 0H1.75a.559.559 0 0 0-.409.175.561.561 0 0 0-.175.41v7c.002.532.105 1.06.305 1.554.189.488.444.948.756 1.368.322.42.682.81 1.076 1.163.365.335.75.649 1.152.939.35.248.718.483 1.103.706.385.222.656.372.815.45.16.08.29.141.386.182A.53.53 0 0 0 7 14a.509.509 0 0 0 .238-.055c.098-.043.225-.104.387-.182.162-.079.438-.23.816-.45.378-.222.75-.459 1.102-.707.403-.29.788-.604 1.154-.939a8.435 8.435 0 0 0 1.076-1.163c.312-.42.567-.88.757-1.367a4.19 4.19 0 0 0 .304-1.555v-7a.55.55 0 0 0-.174-.407Z"
        fill="#175DDC"
      />
      <path
        d="M7 12.365s4.306-2.18 4.306-4.717V1.5H7v10.865Z"
        fill="#fff"
      />
      <circle
        cx="12.889"
        cy="12.889"
        fill="#F8F9FA"
        r="4.889"
      />
      <path
        d="M11.26 11.717h2.37v-.848c0-.313-.116-.58-.348-.8a1.17 1.17 0 0 0-.838-.332c-.327 0-.606.11-.838.332a1.066 1.066 0 0 0-.347.8v.848Zm3.851.424v2.546a.4.4 0 0 1-.13.3.44.44 0 0 1-.314.124h-4.445a.44.44 0 0 1-.315-.124.4.4 0 0 1-.13-.3V12.14a.4.4 0 0 1 .13-.3.44.44 0 0 1 .315-.124h.148v-.848c0-.542.204-1.008.612-1.397a2.044 2.044 0 0 1 1.462-.583c.568 0 1.056.194 1.463.583.408.39.611.855.611 1.397v.848h.149a.44.44 0 0 1 .315.124.4.4 0 0 1 .13.3Z"
        fill="#555"
      />
    </g>
    <defs>
      <clippath
        id="a"
      >
        <rect
          fill="#fff"
          height="16"
          rx="2"
          width="16"
        />
      </clippath>
    </defs>
  </svg>
</button>
`;

exports[`AutofillInlineMenuButton initAutofillInlineMenuButton creates the button element with the normal icon when the user's auth status is Unlocked  1`] = `
<button
  aria-label="toggleBitwardenVaultOverlay"
  class="inline-menu-button"
  tabindex="-1"
  type="button"
>
  <svg
    aria-hidden="true"
    class="inline-menu-button-svg-icon logo-icon"
    fill="none"
    height="14"
    viewBox="0 0 14 14"
    width="14"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M12.66.175A.566.566 0 0 0 12.25 0H1.75a.559.559 0 0 0-.409.175.561.561 0 0 0-.175.41v7c.002.532.105 1.06.305 1.554.189.488.444.948.756 1.368.322.42.682.81 1.076 1.163.365.335.75.649 1.152.939.35.248.718.483 1.103.706.385.222.656.372.815.45.16.08.29.141.386.182A.53.53 0 0 0 7 14a.509.509 0 0 0 .238-.055c.098-.043.225-.104.387-.182.162-.079.438-.23.816-.45.378-.222.75-.459 1.102-.707.403-.29.788-.604 1.154-.939a8.435 8.435 0 0 0 1.076-1.163c.312-.42.567-.88.757-1.367a4.19 4.19 0 0 0 .304-1.555v-7a.55.55 0 0 0-.174-.407Z"
      fill="#175DDC"
    />
    <path
      d="M7 12.365s4.306-2.18 4.306-4.717V1.5H7v10.865Z"
      fill="#fff"
    />
  </svg>
</button>
`;
